File: 486-predict-the-winner/486-predict-the-winner.java
class Solution {
    private int total = 0;
    private static int rec(int s , int e , int [] nums , 
                                            HashMap<Pair<Integer, Integer>, Integer> memo){
        Pair<Integer, Integer> cur = new Pair(s, e);
        if (memo.containsKey(cur)){
            return memo.get(cur);
        }
        if (s == e){
            return nums[s];
             }
        int a = nums[s] - rec(s + 1,  e, nums , memo);
        int b = nums[e] - rec(s, e - 1, nums,  memo);
        memo.put(cur,Math.max(a, b));
        return memo.get(cur);
    }
    public boolean PredictTheWinner(int[] nums) {
        HashMap<Pair<Integer, Integer>, Integer> memo = 
                new HashMap<Pair<Integer, Integer>, Integer>();
        return rec(0, nums.length - 1, nums , memo) >= 0;
    }
}

File: 486-predict-the-winner/486-predict-the-winner.py
class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        @cache
        def dp(i, n, is_plr_1):
            if i == n - 1:
                return nums[i]
            ans = 0
            if is_plr_1:
                ans = max(nums[i] + dp(i + 1, n, False), nums[n - 1] + dp(i , n - 1, False))
            else:
                ans = min(dp(i + 1, n, True) - nums[i], dp(i, n - 1, True) - nums[n - 1])
            return ans
        return (dp(0, len(nums), True)) >= 0
                
                    
        
        
        